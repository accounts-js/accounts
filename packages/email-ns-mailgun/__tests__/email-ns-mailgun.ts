import EmailNSMailgun from '../src';

jest.mock('mailgun-js', () => ({
  default: () => ({})
}))


const mailgunSendSpy = jest.fn(email => email)
const passwordEnrollSpy = jest.fn(()=>true)

const mailgun = {
  messages: ()=>({
    send: mailgunSendSpy
  })
}

const passwordPlugin = {
  name: 'password',
  enroll: passwordEnrollSpy
}


const config = {
  mailgun,
  plugins: [passwordPlugin]
}

const emailNSMailgun = new EmailNSMailgun(config)

describe('EmailNSMailgun', () => {

  describe('validateConfiguration', () => {
    
    it('should throw when no configuration object provided', () => {
      expect(()=> new EmailNSMailgun()).toThrowErrorMatchingSnapshot()
    })

    it('should throw when no valid mailgun config provided', () => {
      expect(()=> new EmailNSMailgun({ domain: true })).toThrowErrorMatchingSnapshot()
      expect(()=> new EmailNSMailgun({ apiKey: true })).toThrowErrorMatchingSnapshot()
    })

    it('should throw when no plugins provided', () => {
      expect(()=> new EmailNSMailgun({ domain: true, apiKey: true })).toThrowErrorMatchingSnapshot()
      expect(()=> new EmailNSMailgun({ mailgun: true })).toThrowErrorMatchingSnapshot()
    })

    it('should not throw when valid mailgun config provided', () => {
      expect(()=> new EmailNSMailgun({ domain: true, apiKey: true, plugins: [] })).not.toThrow()
      expect(()=> new EmailNSMailgun({ mailgun: true, plugins: [] })).not.toThrow()
    })

  })

  describe('constructor',() => {
    

    it('should provide a default from value', () => {
      expect(typeof emailNSMailgun.from).toBe('string')
    })

    it('should assign the from value of config', () => {
      expect(new EmailNSMailgun({ ...config, from: 'test' }).from).toBe('test')
    })

  })

  describe('send',() => {

    it('should throw if plugin not found', () => {
      emailNSMailgun.send('test')
      expect(mailgunSendSpy).toBeCalled()
    })
    
  })


  describe('notify',() => {
    

    it('should throw if plugin not found', () => {
      expect(()=>emailNSMailgun.notify('oauth')).toThrowErrorMatchingSnapshot()
    })

    it('should throw if action not found on plugin', () => {
      expect(()=>emailNSMailgun.notify('password', 'action')).toThrowErrorMatchingSnapshot()
    })

    it('should call the action if found on plugin', () => {
      emailNSMailgun.notify('password', 'enroll', { test: true })
      expect(passwordEnrollSpy).toBeCalled()
    })
    
  })

})
