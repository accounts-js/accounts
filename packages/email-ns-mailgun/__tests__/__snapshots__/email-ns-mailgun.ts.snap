// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailNSMailgun notify should throw if action not found on plugin 1`] = `"[ Accounts - EmailNotificationServiceMailgun ] notify : The action \\"action\\" is not found on plugin \\"password\\""`;

exports[`EmailNSMailgun notify should throw if plugin not found 1`] = `"[ Accounts - EmailNotificationServiceMailgun ] notify : The plugin \\"oauth\\" is not registered"`;

exports[`EmailNSMailgun validateConfiguration should throw when no configuration object provided 1`] = `"[ Accounts - EmailNotificationServiceMailgun ] configuration : A configuration object is required"`;

exports[`EmailNSMailgun validateConfiguration should throw when no plugins provided 1`] = `"[ Accounts - EmailNotificationServiceMailgun ] configuration : The configuration object should provide at least one EmailNotificationPlugin"`;

exports[`EmailNSMailgun validateConfiguration should throw when no plugins provided 2`] = `"[ Accounts - EmailNotificationServiceMailgun ] configuration : The configuration object should provide at least one EmailNotificationPlugin"`;

exports[`EmailNSMailgun validateConfiguration should throw when no valid mailgun config provided 1`] = `"[ Accounts - EmailNotificationServiceMailgun ] configuration : The configuration object should either provide a mailgun-js client or an apiKey and a domain"`;

exports[`EmailNSMailgun validateConfiguration should throw when no valid mailgun config provided 2`] = `"[ Accounts - EmailNotificationServiceMailgun ] configuration : The configuration object should either provide a mailgun-js client or an apiKey and a domain"`;
